pkg/apis/core/validation/validation.go

```diff
@@ -3075,6 +3075,39 @@ func validateContainerCommon(ctr *core.Container, volumes map[string]core.Volume
 	allErrs = append(allErrs, validatePullPolicy(ctr.ImagePullPolicy, path.Child("imagePullPolicy"))...)
 	allErrs = append(allErrs, ValidateResourceRequirements(&ctr.Resources, path.Child("resources"), opts)...)
 	allErrs = append(allErrs, ValidateSecurityContext(ctr.SecurityContext, path.Child("securityContext"))...)
+	return allErrs
+}
+
+func validateHostUsers(spec *core.PodSpec, volumes map[string]core.VolumeSource, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+		if spec.SecurityContext != nil && spec.SecurityContext.HostUsers != nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("hostUsers"), "field not supported. Enable the feature gate UserNamespacesSupport"))
+		}
+		return allErrs
+	}
+
+	// Only make the following checks if hostUsers is false (otherwise, the container uses the
+	// same userns as the host, and so there isn't anything to check).
+	if spec.SecurityContext == nil || spec.SecurityContext.HostUsers == nil || *spec.SecurityContext.HostUsers == true {
+		return allErrs
+	}
+
```

Comment:
are there any other fields in the pod spec that cannot be set in combination with hostUsers=false? (privileged, hostNetwork/hostIPC/hostPID, windows.hostProcess, fsUser, fsGroup, runAsUser, etc)?

Reply:
The host namespaces (Network/IPC/PID) should be disallowed.

Reply:
privileged, I think should be allowed, and documented as having full caps in the new user namespace, and no caps in the parent/owner namespace (e.g. root or other parent)

Reply:
similarly from the kep .. "Please note that a mapping is a bijection of host IDs to container IDs. The settings RunAsUser, RunAsGroup, runAsNonRoot in pod.spec.containers.SecurityContext will refer to the ID used inside the container. Therefore, the host IDs the kubelet choses for the pod are not important for the end user changing the pod spec."

Reply:
> The host namespaces (Network/IPC/PID) should be disallowed.

could we discuss such detail it in the KEP first?

The kernel doesn't block that, is there any reason why we should add this additional limitation?

Reply:
I am worried about permission edge cases when we mix host access for some namespaces with a non-associated user namespace for those.

Reply:
@mrunalp those are implementation restrictions that some runtimes have and some don't (and we even talked about removing some of runc with some other community members).

IMHO, here we should not limit things that in some runtimes will work (and in some others might work in the future). If the combination of userns with hostNetwork doesn't work in runc and the user is using runc, then we show an error at startup.

I think that is the correct behavior

Reply:
In any case, I think modify the validation later (i.e. not a blocker). Either if we keep it as it is, to later add more things, or if we add more things now and later remove those extra checks. So, we can discuss which is the perfect validation afterwards, IMHO. 

If you think we should change this for merging, let me know exactly how you want this validation and I can update it :)

Reply:
I wouldn't block on it for alpha but let's finalize that decision in a KEP update.

Reply:
> In any case, I think modify the validation later (i.e. not a blocker)

When making additions to stable APIs, anything we merge and release should have a well understood surface area and path to GA. Merging API changes first and deferring thinking about how the new field interacts with existing fields and the validation required doesn't seem like a good idea to me.

Reply:
@liggitt ok, then let's discuss it now. My thoughts were expressed here already: https://github.com/kubernetes/kubernetes/pull/111090#discussion_r934668374

What do you think?

Reply:
@liggitt @mrunalp @mikebrow ?

Reply:
thinking out loud.. 
hostUsers={"restricted","userns","hostns"} where hostns is default, restricted creates a userns for the pod and disallows Host Network/IPC/PID, and userns creates a userns for the pod and does not disallow Host Network/IPC/PID

Reply:
would also be fine with adding another bool restrictHostNamespacesToUserNS/allowHostNamespacesWithUserNS or a state set userNSrestrictions (optional) {"network", "IPC", "PID"} userNSallows {"network", "IPC", "PID"}

Reply:
A part of the problem here is there has been a long running hope/promise of userns /pod solving breakouts.. and here we are discussing available host namespaces, necessary for certain types of pods, possibly in conflict with that promise.  As Mrunal states permission edge cases are a concern.

Reply:
@mikebrow lot of thoughts about that:
 1. I don't think that has enough merits to break the consistency with ALL the other host namespaces. All are bool, I don't think we have enough reasons to do that.
 2. Why would you want that in the pod spec? I can't understand. If you don't want a pod using hostNetwork use PSS, OPA, etc. We don't need to add a knob to allow/disallow any possible combination of all the options. Why is this special?
 3. The only edge cases that I'm aware of is that some _implementations_ of low level container runtimes don't allow to mix them (like runc) and some do, like crun. What edge cases do you have in mind?

Reply:
Also, can you please explain what are your concerns regarding keeping the validation as it is?

I've expressed why I think this is correct, and no answer in this thread mentioned why not: https://github.com/kubernetes/kubernetes/pull/111090#discussion_r934668374

I took time to explain why I think this is correct, it would be nice to get an answer. Specially if this is considered a blocker.

Reply:
> @liggitt ok, then let's discuss it now. My thoughts were expressed here already: https://github.com/kubernetes/kubernetes/pull/111090#discussion_r934668374
>
> What do you think?

I don't know the domain well enough to have opinions on the interactions of this field with other pod spec fields, so I'll defer to the node approvers on what the validation restrictions should be.

My point is that those interactions need to be decided before merging/releasing this. As our API surface grows, making sure additions have sensible interactions with existing API surface is one of the primary considerations of new development, not something we should defer.

Reply:
IMHO this is sensible. 

@mrunalp thoughts? Please see my explanation on why I think that in https://github.com/kubernetes/kubernetes/pull/111090#discussion_r934668374

Reply:
The capabilities are tied to the user namespace. So, if we mix non host usernamespace with other host namespaces, the container will run into permission issues, so my preference is that we restrict it.

Reply:
Sure, as we talked on slack, I'll restrict this.

Reply:
userns + hostNetwork:  This seems reasonable and I think it is well-defined.

userns + hostIPC:  What UID/GID is used when accessing shared memory or semaphores?  Is it the in-pod UID?  Does it even work?  IOW, assume I have a pod running as UID 6 internally, mapped to 1000006 in the root, and using hostIPC.  That pod creates a shared memory segment.  Now I run another pod, also hostIP, also as UID 6 internally, but mapped to 2000006 in the root.  Can it access that SHM?

userns + hostPID:  This seems reasonable and I think it is well-defined?

My default position on things I don't fully understand is to leave them out - we can always enable it later, but we can't turn it off.  So, if we are unsure of the semantics and/or usefulness, let's just disallow it for now.

Agree or disagree?

Reply:
Yes, I think we are all on the same page. Disallow it for now.

The long answer would be:
At the theoretical POV you mention, all those things are probably well-defined. There is only one thing that happens (works or not, but it is always the same).

That is not the point, but the point is:
 1. userns + hostNetwork: this works in some implementations of low level container runtimes and not with others. It is not a limitation in the Linux kernel, just on how the runc/crun are implemented.
 1. userns + hostNetwork + CAP_NET_ADMIN: this won't give you capabilities on the host network (this is a limitation imposed by the linux kernel on how userns work). However, this _might_ be confusing for users.

Something similar happens with other namespaces too.

I was arguing that we shouldn't add artificial limitations here (1 works with some runtimes, why limit them? And 2 might be confusing if you don't know the details, but those are explained and documented).

However, we agreed with Mrunal on what you are saying: let's limit them all for now (except probably shareProcessNamespaces) and we can enable them later.

I think we are all on the same page here :)

Reply:
Whe userns are in use, restricted the use of: HostNetwork, HostPID, HostIPC, ShareProcessNamespace and privileged containers.

If someone want to enable any of this options, we can discuss it later, to do in another release.

The whole point of doing this, is that we can easily enable them later, so don't worry to discuss this now ;)